docs/csharp/roslyn-sdk/tutorials/how-to-write-csharp-analyzer-code-fix.md

```diff
@@ -0,0 +1,615 @@
+* Every Diagnostic Analyzer must derive from the <xref:Microsoft.CodeAnalysis.Diagnostics.DiagnosticAnalyzer> class.
+
+The analyzer project template also creates a test project template for you. Look at the **MakeConstAnalyzerUnitTest.cs** file in the test project. It contains two tests: one that analyzes the empty string, and one that analyzes a snippet of code with a class whose name contains lower case letters. The first shows the general format for a test where your analyzer does not find code that it should change. The second shows the general format for a test where your analyzer finds code that it modifies.
+
+Throughout this tutorial, you'll use a combination of running your analyzer in a second copy of Visual Studio, and adding new unit tests that validate your analyzer and code fix. This combination provides a way to explore your analyzer's capabilities as you build it, and create quality tests to prevent regressions and validate actions quickly. This practice is often a good way to proceed with your analyzers: experiment in Visual Studio, and then write tests as you learn more about the code to spot and change.
+
+## Create the analyzer
+
+The template shows the basic features that are part of any analyzer:
+
+1. Register actions. The actions represent code changes that should trigger your analyzer to examine code for violations. When Visual Studio detects code edits that match a registered action, it calls your analyzer's registered method.
+1. Create diagnostics. When your analyzer detects a violation, it creates a diagnostic object that Visual Studio uses to notify the user of the violation.
+
+You register actions in your override of <xref:Microsoft.CodeAnalysis.Diagnostics.DiagnosticAnalyzer.Initialize(Microsoft.CodeAnalysis.Diagnostics.AnalysisContext)?displayProperty=nameWithType> method. The actions your register create the diagnostics. In this tutorial, you'll visit **syntax nodes** looking for local declarations, and see which of those have constant values.
+
+### Create the analyzer structure
+
+Open **MakeConstAnalyzer.cs** in Visual Studio. Change the registered action from one that acts on symbols to one that acts on syntax. In the `MakeConstAnalyzerAnalyzer.Initializer` method, find the line that registers the action on symbols:
+
+```csharp
+context.RegisterSymbolAction(AnalyzeSymbol, SymbolKind.NamedType);
+```
+
+Replace it with the following line:
+
+[!code-csharp[Register the node action](~/samples/csharp/roslyn-sdk/Tutorials/MakeConst/MakeConst/MakeConst/MakeConstAnalyzer.cs#RegisterNodeAction "Register a node action")]
+
+After that change, you can delete the `AnalyzeSymbol` method. This analyzer examines <xref:Microsoft.CodeAnalysis.CSharp.SyntaxKind.LocalDeclarationStatement?displayProperty=nameWithType>, not <xref:Microsoft.CodeAnalysis.SymbolKind.NamedType?displayProperty=nameWithType> statements. Then, put your cursor on the `AnalyzeNode` method, press **Ctrl+.** and generate the `AnalyzeNode` method.
```

Comment:
Should you use keyboard shortcuts versus mouse instructions? The user is told to move the cursor, not really indicating keyboard or mouse input, but in general, cursor indicates mouse movement. So maybe a **right-click > Quick-actions and refactoring > Generate method** and then put in `(or press CTRL + .)`

OR just have them copy/paste the method skeleton to replace the `AnalyzeSymbol` method.

--EDIT--
OK after the next paragraph, I see that you're trying to show them that pasting in the code causes that pasted code to be evaluated and the existing analyzer saw that there was no method named `AnalyzeNode`. Maybe these details should be right after the instruction to paste instead of the after the instruction to delete the old method?